File: mobile/ios/OpenU60/Core/Networking/ClashDirectClient.swift
import Foundation

@MainActor
final class ClashDirectClient {
    private static let defaultSecret = "123456"
    private static let defaultPort = 7788

    private let session: URLSession

    init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 5
        config.timeoutIntervalForResource = 10
        session = URLSession(configuration: config)
    }

    func getJSON(agentBaseURL: String, path: String) async throws -> [String: Any] {
        let data = try await request(agentBaseURL: agentBaseURL, method: "GET", path: path, body: nil)
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw AgentError.decodingError("预期为 JSON 对象")
        }
        return json
    }

    func putJSON(agentBaseURL: String, path: String, body: [String: Any]) async throws -> [String: Any] {
        let bodyData = try JSONSerialization.data(withJSONObject: body)
        let data = try await request(agentBaseURL: agentBaseURL, method: "PUT", path: path, body: bodyData)
        if data.isEmpty { return [:] }
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw AgentError.decodingError("预期为 JSON 对象")
        }
        return json
    }

    func patchJSON(agentBaseURL: String, path: String, body: [String: Any]) async throws -> [String: Any] {
        let bodyData = try JSONSerialization.data(withJSONObject: body)
        let data = try await request(agentBaseURL: agentBaseURL, method: "PATCH", path: path, body: bodyData)
        if data.isEmpty { return [:] }
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw AgentError.decodingError("预期为 JSON 对象")
        }
        return json
    }

    private func request(agentBaseURL: String, method: String, path: String, body: Data?) async throws -> Data {
        guard let agentURL = URL(string: agentBaseURL),
              let host = agentURL.host else {
            throw AgentError.serverUnreachable
        }
        var components = URLComponents()
        components.scheme = agentURL.scheme ?? "http"
        components.host = host
        components.port = Self.defaultPort
        components.percentEncodedPath = path

        guard let url = components.url else {
            throw AgentError.serverUnreachable
        }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("Bearer \(Self.defaultSecret)", forHTTPHeaderField: "Authorization")
        if let body {
            request.httpBody = body
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch let error as URLError where error.code == .timedOut {
            throw AgentError.timeout
        } catch let error as URLError where error.code == .cannotConnectToHost || error.code == .notConnectedToInternet {
            throw AgentError.serverError("无法连接 Clash 控制器，请确认 7788 端口可访问")
        } catch {
            throw AgentError.networkError(error)
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            throw AgentError.serverUnreachable
        }

        switch httpResponse.statusCode {
        case 200...299:
            return data
        case 401:
            throw AgentError.serverError("Clash API 鉴权失败，请检查 secret 是否仍为默认 123456")
        default:
            let message = String(data: data, encoding: .utf8) ?? "HTTP \(httpResponse.statusCode)"
            throw AgentError.serverError(message)
        }
    }
}

File: mobile/android/OpenU60/app/src/main/java/com/openu60/core/network/ClashDirectClient.kt
package com.openu60.core.network

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import java.net.SocketTimeoutException
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ClashDirectClient @Inject constructor() {

    companion object {
        private val JSON_MEDIA_TYPE = "application/json".toMediaType()
        private const val DEFAULT_SECRET = "123456"
        private const val DEFAULT_PORT = 7788
    }

    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(5, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .writeTimeout(5, TimeUnit.SECONDS)
        .build()

    suspend fun getJSON(agentBaseURL: String, path: String): Map<String, Any?> {
        val data = request(agentBaseURL, "GET", path, null)
        val element = Json.parseToJsonElement(data)
        return element.jsonObject.toMap()
    }

    suspend fun putJSON(agentBaseURL: String, path: String, body: Map<String, Any?>): Map<String, Any?> {
        val data = request(agentBaseURL, "PUT", path, mapToJsonString(body))
        return if (data.isBlank()) emptyMap() else Json.parseToJsonElement(data).jsonObject.toMap()
    }

    suspend fun patchJSON(agentBaseURL: String, path: String, body: Map<String, Any?>): Map<String, Any?> {
        val data = request(agentBaseURL, "PATCH", path, mapToJsonString(body))
        return if (data.isBlank()) emptyMap() else Json.parseToJsonElement(data).jsonObject.toMap()
    }

    private suspend fun request(
        agentBaseURL: String,
        method: String,
        path: String,
        body: String?,
    ): String = withContext(Dispatchers.IO) {
        val url = controllerBaseURL(agentBaseURL) + path
        val requestBody = body?.toRequestBody(JSON_MEDIA_TYPE)
        val builder = Request.Builder()
            .url(url)
            .header("Authorization", "Bearer $DEFAULT_SECRET")

        when (method) {
            "GET" -> builder.get()
            "PUT" -> builder.put(requestBody ?: "".toRequestBody(JSON_MEDIA_TYPE))
            "PATCH" -> builder.patch(requestBody ?: "".toRequestBody(JSON_MEDIA_TYPE))
            else -> error("Unsupported method $method")
        }

        if (body != null) {
            builder.header("Content-Type", "application/json")
        }

        try {
            val response = httpClient.newCall(builder.build()).execute()
            val responseBody = response.body?.string() ?: ""
            when (response.code) {
                in 200..299 -> responseBody
                401 -> throw IllegalStateException("Clash API 鉴权失败，请检查 secret 是否仍为默认 123456")
                else -> throw IllegalStateException(responseBody.ifBlank { "Clash HTTP ${response.code}" })
            }
        } catch (e: SocketTimeoutException) {
            throw IllegalStateException("连接 Clash 控制器超时")
        } catch (e: java.net.ConnectException) {
            throw IllegalStateException("无法连接 Clash 控制器，请确认 7788 端口可访问")
        }
    }

    private fun controllerBaseURL(agentBaseURL: String): String {
        val parsed = agentBaseURL.toHttpUrlOrNull()
        val scheme = parsed?.scheme ?: "http"
        val host = parsed?.host ?: "192.168.0.1"
        return "$scheme://$host:$DEFAULT_PORT"
    }

    private fun mapToJsonString(map: Map<String, Any?>): String {
        return buildJsonObject {
            for ((key, value) in map) {
                putAny(key, value)
            }
        }.toString()
    }
}
